### PR TITLE
[DEM-NNN] Add 1/f noise analysis

### PR DESCRIPTION
**Pull request by peendebak**
* Add calculation of PSD
* Add 1/f noise analysis
* Add example notebook

### CODE REVIEW

File: src/qtt/algorithms/noise_analysis.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    else:
+        ax = plt.gca()
+    ax.loglog(frequencies, psd, label=label, color='navy')
+    plt.xlabel(f'Frequency ({frequency_unit})')
+    plt.ylabel(f'Power spectral density ({signal_unit}$^2$/{frequency_unit})')
+
+
+def generate_pink_noise(sample_rate: float, number_of_samples: int, A: float = 1, alpha: float = 1):
+    """ Generate sampled data with 1/f noise
+
```
Comment (sldesnoo-Delft) [inline src/qtt/algorithms/noise_analysis.py:67]: Seems to be based on:
    Timmer, J. and Koenig, M.:
    On generating power law noise.
    Astron. Astrophys. 300, 707-710 (1995)

Comment (peendebak) [reply src/qtt/algorithms/noise_analysis.py:67]: Fair point. I added a reference

File: src/qtt/algorithms/noise_analysis.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        A: Parameter of noise model
+        alpha: Parameter of noise model
+
+    Returns:
+        Array with sampled data
+    """
+    frequencies = rfftfreq(number_of_samples, 1 / sample_rate)
+
+    s_scale = frequencies
+    s_scale[0] = s_scale[1]
```
[the remark this answers is not in the document]
Comment (peendebak) [reply src/qtt/algorithms/noise_analysis.py:80]: The DC component is set to zero a few lines below, this sets up the scaling. Frequency zero is a bit tricky as for 1/f noise we have a singularity at f=0.

File: src/qtt/algorithms/noise_analysis.py [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        plt.clf()
+        ax = Fig.add_subplot(111)
+    else:
+        ax = plt.gca()
+    ax.loglog(frequencies, psd, label=label, color='navy')
+    plt.xlabel(f'Frequency ({frequency_unit})')
+    plt.ylabel(f'Power spectral density ({signal_unit}$^2$/{frequency_unit})')
+
+
+def generate_pink_noise(sample_rate: float, number_of_samples: int, A: float = 1, alpha: float = 1):
```
[the remark this answers is not in the document]
Comment (peendebak) [reply src/qtt/algorithms/noise_analysis.py:65]: I updated the documentation to add the names for the noise variants. For the name I picked `generate_powerlaw_noise`

File: src/qtt/algorithms/noise_analysis.py [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+
+def outlier_detection(data: np.ndarray, threshold: Optional[float] = None,
+                      percentile: float = 90):
+    """ Detect outliers in data using a threshold
+
+    Args:
+        data: Data from which to determine the outliers
+        threshold: Threshold to use for outlier detection. If None, then use the percentile argument to automatically determine the threshold
+        percentile: Determine the threshold by setting it the the percentile specified
```
[the remark this answers is not in the document]
Comment (peendebak) [reply src/qtt/algorithms/noise_analysis.py:102]: 10% is indeed a lot, but we do not know beforehand which percentage will be outliers. For data with very few outliers (e.g. < 1%), thresholding at 10% will often result in a very reasonable fit.

File: src/qtt/algorithms/noise_analysis.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+        frequency: Independent variable
+        A: Scalar factor of the model
+        alpha: Exponent of the model
+    Returns:
+        Calculated value of the model
+    """
+    return A / frequency**alpha
+
+
+def calculate_psd_welch(signal: np.ndarray, sample_rate: float, nperseg: Optional[int] = None) -> Tuple[np.ndarray, np.ndarray]:
```
[the remark this answers is not in the document]
Comment (peendebak) [reply src/qtt/algorithms/noise_analysis.py:36]: @sldesnoo-Delft Added value is that we have type annotations and the naming of the arguments is consistent with the other methods in the module (but these are minor advantages). 

@QFer Do you have a strong opinion here?

[the remark this answers is not in the document]
Comment (QFer) [reply src/qtt/algorithms/noise_analysis.py:36]: I think that when the doc-string is adding value and the functions around noise analysis are bundled this way it is oke to do.

**Comment by eendebakpt** [comment]
Closing as there is nobody to review.

**Pull request closed without merging**